### PR TITLE
Added VisibleFrom datetime field in contest

### PR DESCRIPTION
**Pull request by shsimeonova**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/205

### CODE REVIEW

**Review by mishogv — APPROVED** [review]

File: Open Judge System/Data/OJS.Data.Models/Contest.cs [hunk 1]
```diff
@@ -102,12 +104,31 @@ public class Contest : DeletableEntity, IValidatableObject, IOrderable
...
+                    return false;
+                }
+
+                if (this.IsVisible || (this.VisibleFrom != null && this.VisibleFrom <= DateTime.Now))
+                {
+                    return true;
+                }
+
+                return this.IsVisible;
+            }
```
Comment (Minkov) [inline Open Judge System/Data/OJS.Data.Models/Contest.cs:123]: Isn't this the same:
```
return this.IsVisible || this.VisibleFrom < DateTime.Now;
```

File: Open Judge System/Services/OJS.Services/Data/Contests/ContestsDataService.cs [hunk 2]
```diff
@@ -62,7 +62,7 @@ public class ContestsDataService : IContestsDataService
 
         public IQueryable<Contest> GetAllVisible() =>
             this.GetAll()
-                .Where(c => c.IsVisible);
+                .Where(c => c.IsVisible || (c.VisibleFrom != null && c.VisibleFrom <= DateTime.Now));
```
Comment (Minkov) [inline Open Judge System/Services/OJS.Services/Data/Contests/ContestsDataService.cs:65]: Why don't you use `Visible` here?

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Contests/ContestViewModel.cs [hunk 3]
```diff
@@ -116,6 +119,24 @@ public string CategoryName
...
+                    return false;
+                }
+
+                if (this.IsVisible || (this.VisibleFrom != null && this.VisibleFrom <= DateTime.Now))
+                {
+                    return true;
+                }
+
+                return this.IsVisible;
+            }
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Contests/ContestViewModel.cs:137]: Isn't this the same check as above?

**Review by Minkov — CHANGES_REQUESTED** [review]

File: Open Judge System/Data/OJS.Data.Models/Contest.cs (hunk 1 above)
Comment (shsimeonova) [reply Open Judge System/Data/OJS.Data.Models/Contest.cs]: Yes it is. I've edited it.

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Contests/ContestViewModel.cs (hunk 3 above)
Comment (shsimeonova) [reply Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Contests/ContestViewModel.cs]: Yes, but i think it cannot be unified to use only the field and logic from Contest.cs as the view model does not inherit the Contest class, thus why i have to copy the logic here, same as CanBeCompeted and CanBePracticed fields. 

I've thought about moving this in a method in ContestExtensions, but it has to be made Common (we currently have one in OJS.Web project).

If you have any ideas how we can change this, please advise.

**Review by Minkov — APPROVED** [review]

**Review by Minkov — DISMISSED** [review]

**Review by mishogv — DISMISSED** [review]

**Review by mishogv — APPROVED** [review]

**Pull request merged**